core/rawdb/chain_iterator.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package rawdb
+
+import (
+	"errors"
+	"math"
+	"runtime"
+	"sync/atomic"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/prque"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type (
+	prepareCallback func(*types.Block)              // Callback for custom concurrent pre-computations on block data
+	actionCallback  func(ethdb.Batch, *types.Block) // Callback for custom sequential operations on block data.
+)
+
+// iterateCanonicalChain iterates the specified range of canonical blocks and applies
+// the given action callback. Note, both for forward and backward iteration, the range
+// is [from, to).
+func iterateCanonicalChain(db ethdb.Database, from uint64, to uint64, prepare prepareCallback, action actionCallback, reverse bool, progMsg, doneMsg string) error {
+	// Short circuit if the action is nil or if the range is invalid
+	if action == nil {
+		return nil
+	}
+	if from >= to {
+		return nil
+	}
+	// Spawn multi-routines, iterate over the specified blocks and invoke prepare
+	// callback concurrently.
+	var (
+		number  int64
+		results = make(chan *types.Block, 4*runtime.NumCPU())
+	)
+	if !reverse {
+		number = int64(from - 1)
+	} else {
+		number = int64(to)
+	}
+	abort := make(chan struct{})
+	defer close(abort)
+
+	threads := to - from
+	if cpus := runtime.NumCPU(); threads > uint64(cpus) {
+		threads = uint64(cpus)
+	}
+	for i := 0; i < int(threads); i++ {
+		go func() {
+			for {
+				// Fetch the next task number, terminating if everything's done
+				var n int64
+				if !reverse {
+					n = atomic.AddInt64(&number, 1)
+					if n >= int64(to) {
+						return
+					}
+				} else {
+					n = atomic.AddInt64(&number, -1)
+					if n < int64(from) {
+						return
+					}
+				}
+				block := ReadBlock(db, ReadCanonicalHash(db, uint64(n)), uint64(n))
```

Comment:
The way this is done is simple, and that's good, I guess. It just irks me a bit that given the format of `ancients`, it has a pretty large overhead. So this is what happens, assuming that things aren't cached. 
1. Read hash from freezer (tiny 32 byte read)
2. Read header rlp from freezer. Decode header RLP into header.
3. Read body rlp from freezer. Decode body rlp into`types.Body`.
4. Construct new Block with header with body (copying heder, transactoins and uncles)
5. And then call prepare on it

Whereas, for everything that's in `ancient`s, it would basically suffice to: 
1. Read body rlp from freezer. Decode into `types.Body`. 
2. call prepare on the body.Transactions

That would save us 2 disk accesses and a lot of `memcpy`. Maybe we could have a method `ReadCanonicalTransactions(blocknumber)`, so we know that we can pull straight from ancients ? I don't now what the best solution is, thinking aloud here...